### PR TITLE
fix(consult): changes for EPDN consult transfer

### PR DESCRIPTION
**Pull request by Shreyas281299**
# COMPLETES #N/A (EP-DN consult transfer defect fix)

## This pull request addresses

EP-DN consult transfer regressions where task state and UI controls could drift after transfer-related event ordering differences. In affected flows, agents could see incorrect controls (consult/transfer/end visibility), incorrect consulted-role inference, or miss wrap-up progression after transfer.

## by making the following changes

- Added effective voice-state derivation to align UI/operation guards with latest interaction payload when XState snapshot lags.
- Hardened transfer payload normalization in task lifecycle handling (`wrapUpRequired` and `isConsulted` inference), including owner-aware EP-DN consulted detection.
- Updated state-machine guards/transitions for transfer and wrap-up handling (`TRANSFER_SUCCESS` routing and `WRAPUP_COMPLETE` completion handling).
- Improved consult leg detection and consult-control computation for payloads with partial consult metadata.
- Added/expanded unit coverage for TaskUtils, state machine transitions, effective voice state mapping, and UI control behavior.

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [x] Internal code refactor

## The following scenarios were tested

- Unit tests for `uiControlsComputer` consult-leg visibility and control enablement in EP-DN/partial-payload shapes.
- Unit tests for `TaskStateMachine` transfer/wrap-up lifecycle behavior, including `WRAPUP_COMPLETE` progression.
- Unit tests for `effectiveVoiceTaskState` wrap-up forcing and non-wrap-up fallbacks.
- Unit tests for `TaskUtils` EP-DN consulted-recipient inference.
- Manual validation of consult transfer/blind transfer edge cases (state + controls + wrap-up progression).

## The GAI Coding Policy And Copyright Annotation Best Practices ##
 
- [ ] GAI was not used (or, no additional notation is required)
- [ ] Code was generated entirely by GAI
- [x] GAI was used to create a draft that was subsequently customized or modified
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [x] Tool used for AI assistance (GitHub Copilot / Other - specify)
  - [ ] Github Copilot
  - [x] Other - Please Specify (ChatGPT Codex)
- [x] This PR is related to
  - [ ] Feature
  - [x] Defect fix
  - [ ] Tech Debt
  - [ ] Automation

### I certified that

- [ ] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [ ] I discussed changes with code owners prior to submitting this pull request
- [ ] I have not skipped any automated checks
- [ ] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

**Comment by Shreyas281299** [comment]
### Changeset `8626ee4` — UI consult-leg visibility stabilization

**What changed**
- Added `getIsConsultInProgress()` in `TaskUtils` to derive consult-in-progress from explicit task flag or consult media.
- Updated `uiControlsComputer` to use this derived signal for consult leg visibility and control state.
- Added unit tests in `uiControlsComputer` spec for payloads where consult media is unavailable and consult state must still be inferred.

**Why it changed**
- EP-DN payload ordering can temporarily omit consult media/state details, causing UI controls to show incorrect leg/actions.
- This changeset makes control rendering resilient to those transient payload shapes so users see the expected consult controls consistently.

**Comment by Shreyas281299** [comment]
### Changeset `6440016` — transfer/wrap-up/state-machine hardening for EP-DN consult flows

**What changed**
- Added `effectiveVoiceTaskState` and wired it into UI computation and voice operation guards.
- Improved `TaskManager` transfer payload normalization (`wrapUpRequired` and `isConsulted` inference), including EP-DN owner-aware consulted detection.
- Updated state-machine guards/transitions to route transfer outcomes to `WRAPPING_UP` using per-agent wrap-up checks.
- Added root handling for `WRAPUP_COMPLETE` so delayed event ordering still completes cleanup correctly.
- Added/expanded tests for state machine, effective voice state, TaskUtils, and UI controls.

**Why it changed**
- In consult/blind transfer edge cases, stale or partial backend payloads caused wrong control visibility, incorrect consulted-role inference, and missed wrap-up transitions.
- This changeset aligns machine state + UI controls with the effective interaction state so both transfer recipient and transfer initiator follow the correct lifecycle and cleanup path.

File: packages/@webex/contact-center/src/services/task/TaskManager.ts [hunk 1]
```diff
@@ -426,6 +431,16 @@ export default class TaskManager extends EventEmitter {
 
     const wasConsultedTask = Boolean(task?.data?.isConsulted);
     const computeWrapUpRequired = () => {
+      const pending = message.data.agentsPendingWrapUp;
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/TaskManager.ts:434]: Inline note: this transfer-specific wrap-up derivation prioritizes `agentsPendingWrapUp` and participant `isWrapUp` for the current agent before fallback heuristics. That prevents Agent-2 from staying on active controls when backend transfer payloads don’t set `wrapUpRequired` consistently.

File: packages/@webex/contact-center/src/services/task/state-machine/effectiveVoiceTaskState.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  if (!selfId) return Boolean(taskData.wrapUpRequired);
+
+  const pending = taskData.agentsPendingWrapUp;
+  if (Array.isArray(pending) && pending.length > 0) {
+    return pending.includes(selfId);
+  }
+
+  const participantWrapUp = taskData.interaction?.participants?.[selfId]?.isWrapUp === true;
+  if (participantWrapUp) return true;
+  const participants = taskData.interaction?.participants;
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/state-machine/effectiveVoiceTaskState.ts:20]: Inline note: `shouldForceWrapUpForCurrentAgent()` intentionally treats missing/self-left participant entries as wrap-up signals. In transfer races this makes UI state deterministic (`WRAPPING_UP`) even when machine state or payload flags arrive out of order.

File: packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts [hunk 3]
```diff
@@ -78,6 +78,12 @@ export function getTaskStateMachineConfig(uiControlConfig: UIControlConfig) {
       [TaskEvent.HYDRATE]: {
         actions: ['updateTaskData', 'emitTaskHydrate'],
       },
+      // Wrapped-up can arrive while machine still reflects CONNECTED/CONSULTING due event ordering.
+      // Accept it globally so completion cleanup always runs and task is removed from collection.
+      [TaskEvent.WRAPUP_COMPLETE]: {
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/state-machine/TaskStateMachine.ts:83]: Inline note: this root-level `WRAPUP_COMPLETE` transition is added to handle event-ordering gaps where wrapped-up arrives before state reaches `WRAPPING_UP`; it guarantees completion/cleanup instead of leaving stale task cards.

File: packages/@webex/contact-center/src/services/task/state-machine/uiControlsComputer.ts [hunk 4]
```diff
@@ -108,7 +111,17 @@ function computeVoiceInteractionUIControls(
     mainCallId,
     selfAgentId
   );
+  const consultInProgressForVisibility = getIsConsultInProgress(taskData);
   const conferenceFromBackend = taskData ? getIsConferenceInProgress(taskData) : false;
+  const consultRelationship = interaction?.callProcessingDetails?.relationshipType === 'consult';
+  // Stale relationshipType=consult often remains after consult transfer to a new primary owner.
+  // Only treat it as an active consult session for UI when the interaction is consulting or this
+  // task is the secondary (EP-DN) consult leg — not a warm-transferred main call.
+  const consultRelationshipAffectsConsultingUi = Boolean(
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/state-machine/uiControlsComputer.ts:120]: Inline note: `consultRelationshipAffectsConsultingUi` narrows stale `relationshipType=consult` handling so only true consult contexts get consult controls. This avoids showing consult-only UI to a newly transferred primary owner.

File: packages/@webex/contact-center/test/unit/spec/services/task/state-machine/TaskStateMachine.ts [hunk 5]
```diff
@@ -157,6 +158,18 @@ describe('Task state machine', () => {
       service.send({type: TaskEvent.WRAPUP_COMPLETE});
       expect(service.getSnapshot().value).toBe(TaskState.COMPLETED);
     });
+
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/test/unit/spec/services/task/state-machine/TaskStateMachine.ts:161]: Inline note: this test locks the transfer regression: if wrap-up completion arrives before explicit `WRAPPING_UP`, machine must still reach `COMPLETED` so cleanup/removal happens deterministically.

File: packages/@webex/contact-center/src/services/task/TaskUtils.ts [hunk 6]
```diff
@@ -236,6 +250,26 @@ export const isSecondaryEpDnAgent = (interaction: Interaction): boolean => {
   return interaction.mediaType === 'telephony' && isSecondaryAgent(interaction);
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/TaskUtils.ts:250]: Inline note: owner-aware consulted detection is critical after consult transfer. Secondary EP-DN shape can still be present, but if `interaction.owner` is self, this task is now primary and must not remain in consulted-only UI/state handling.

File: packages/@webex/contact-center/src/services/task/voice/Voice.ts [hunk 7]
```diff
@@ -439,17 +434,15 @@ export default class Voice extends Task implements IVoice {
    * ```
    * */
   public async consult(consultPayload?: ConsultPayload): Promise<TaskResponse> {
-    // Validate consult is allowed
-    const state = this.getStateMachineSnapshot();
+    // Validate consult is allowed (use effective state when machine lags IDLE after transfer)
+    const effectiveState = this.getEffectiveVoiceTaskStateForOperation();
     const canConsult =
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/voice/Voice.ts:439]: Inline note: consult validation here intentionally uses effective state (not raw snapshot) to handle post-transfer ordering where machine can temporarily lag at `IDLE` while payload already reflects an active voice leg.

File: packages/@webex/contact-center/src/services/task/Task.ts [hunk 8]
```diff
@@ -753,6 +776,12 @@ export default abstract class Task extends EventEmitter implements ITask {
         data: wrapupPayload,
       });
 
+      // Some transfer flows can miss or delay AGENT_WRAPPEDUP; complete locally on API success.
```
Comment (Shreyas281299) [inline packages/@webex/contact-center/src/services/task/Task.ts:779]: Inline note: local `WRAPUP_COMPLETE` dispatch on wrapup API success is a reliability guard. It prevents stale wrapup dropdown/task cards when websocket `AGENT_WRAPPEDUP` is delayed or dropped.
